plugin/src/main/java/org/opensearch/ml/utils/SecurityUtils.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.ml.utils;
+
+import static org.opensearch.ml.common.CommonValue.ML_MODEL_GROUP_INDEX;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import lombok.extern.log4j.Log4j2;
+
+import org.apache.commons.lang3.StringUtils;
+import org.opensearch.action.ActionListener;
+import org.opensearch.action.get.GetRequest;
+import org.opensearch.client.Client;
+import org.opensearch.common.util.CollectionUtils;
+import org.opensearch.commons.authuser.User;
+import org.opensearch.index.query.BoolQueryBuilder;
+import org.opensearch.index.query.QueryBuilder;
+import org.opensearch.index.query.QueryBuilders;
+import org.opensearch.index.query.TermsQueryBuilder;
+import org.opensearch.ml.common.MLModelGroup;
+import org.opensearch.ml.common.exception.MLValidationException;
+import org.opensearch.search.builder.SearchSourceBuilder;
+
+import com.google.common.collect.ImmutableList;
+
+@Log4j2
+public class SecurityUtils {
+
+    public static void validateModelGroupAccess(User user, String modelGroupId, Client client, ActionListener<Boolean> listener) {
+        if (modelGroupId == null || isAdmin(user)) {
+            listener.onResponse(Boolean.TRUE);
+            return;
+        }
+
+        List<String> userBackendRoles = user.getBackendRoles();
+
+        GetRequest getModelGroupRequest = new GetRequest(ML_MODEL_GROUP_INDEX).id(modelGroupId);
+        try {
+            client.get(getModelGroupRequest, ActionListener.wrap(r -> {
+                if (r != null && r.isExists()) {
+                    List<String> modelGroupBackendRoles = new ArrayList<>();
+                    String access = null;
+                    Map<String, Object> owner = null;
+                    Map<String, Object> source = r.getSourceAsMap();
+                    if (source.containsKey(MLModelGroup.ACCESS)) {
+                        access = (String) source.get(MLModelGroup.ACCESS);
+                    }
+                    if (source.containsKey(MLModelGroup.BACKEND_ROLES_FIELD) && source.get(MLModelGroup.BACKEND_ROLES_FIELD) != null) {
+                        modelGroupBackendRoles = (List<String>) source.get(MLModelGroup.BACKEND_ROLES_FIELD);
+                    }
+                    if (source.containsKey(MLModelGroup.OWNER)) {
+                        owner = (Map) source.get(MLModelGroup.OWNER);
+                    }
+                    if (access != null && StringUtils.equals(access, MLModelGroup.PUBLIC)) {
+                        listener.onResponse(Boolean.TRUE);
+                    } else if (access != null && StringUtils.equals(access, MLModelGroup.PRIVATE)) {
+                        if (isOwner(owner, user)) {
+                            listener.onResponse(Boolean.TRUE);
+                            return;
+                        }
```

Comment:
Since you're almost parsing all the fields of MLModelGroup, it's better to use the parse XContent approach to create a MLModelGroup like this:
```
XContentParser parser = MLNodeUtils.createXContentParserFromRegistry(NamedXContentRegistry.EMPTY, x.getSourceRef())
                    ) {
                        ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.nextToken(), parser);
                        MLModelGroup mlModelGroup = MLModelGroup.parse(parser);
```

Reply:
Thanks Zan, makes sense. Initially there were only backend_roles field I was using to filter them out. Later had to add more fields.